site/docs/v1/tech/guides/advanced-registration-forms.adoc

```diff
@@ -0,0 +1,540 @@
+
+Any validation failure will prevent the user from moving past the current registration step. The theme controls the location and display of error messages. All validation for advanced registration forms are either browser native or server side. If you'd like to add client side validation, you may inject JavaScript validation libraries and code into your login templates.
+
+=== Form Control
+
+If your field uses a form control with a limited set of options, such as a radio button or select dropdown, the user will be forced to choose from that set of options. 
+
+Form field control options are documented in the link:../apis/form-fields[form field API documentation].
+
+=== Data Type
+
+You can configure a form field to use one of the non-`String` data types. Doing so means the form field will require the user to enter data acceptable to that data type. For instance, if a form field has a data type of `Number`, any non-numeric value will result in an error message.
+
+Form field data type options are thoroughly documented in the link:../apis/form-fields[form field API documentation].
+
+=== The Required Attribute
+
+If a field is configured to be [field]#required#, a valid value must be provided. Otherwise an empty string is a valid value.
+
+=== The Confirm Value Attribute
+
+If a field is configured to have a [field]#Confirm value#, a second input field of the same type and control will be added to the form. This confirmation field will be displayed just below the original field, but the location can be customized by modifying the theme. 
+
+The form will fail validation unless the same value is entered in both fields.
+
+=== Regular Expression Validation
+
+If [field]#Validation# is enabled, a regular expression must be specified. The user input will be matched against the regular expression and validation will fail if it doesn't match. See the https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html[Java Regular Expression documentation] for more information on how to build such a regular expression.
+
+== Special Considerations
+
+=== Adding Required Fields Later
+
+Once you enable self service registration, the authentication flow is:
+
+```
+Authorize -> Complete Registration -> Redirect
+```
+
+Every time a user authenticates using the hosted login pages, FusionAuth ensures their registration is complete.
+
+If you add a required field to the application's registration form after users have registered, the next time one of the users authenticates using the hosted login pages, they'll be sent to the registration form to fill out the required field. The [field]#OAuth complete registration# template will be used in this scenario.
+
+=== Modifying an Existing Form Field
+
+You cannot change the underlying field, control or data type of an existing form field. Other attributes may be modified. 
+
+If you need to change the data type of a field, create a new one. Duplicate the form field and update the form to use the duplicate.
```

Comment:
We could also add a warning here that once you start storing data on users or registrations it is indexed (if using Elasticsearch) and changing data types is bad. 
So you can change the field name if the data type has to change, or you may have to update all users and then re-index to fix the index if you break it.

Reply:
Ah, that's a good point. I modified to reflect that advice.